warehouse/accounts/views.py

```diff
@@ -814,3 +818,52 @@ def edit_profile_button(user, request):
 )
 def profile_public_email(user, request):
     return {"user": user}
+
+
+@view_config(
+    require_methods=["POST"],
+    require_csrf=False,
+    renderer="json",
+    route_name="accounts.github-disclose-token",
+    headers=["GITHUB-PUBLIC-KEY-IDENTIFIER", "GITHUB-PUBLIC-KEY-SIGNATURE"],
+    has_translations=True,  # Not the view itself, but the email it sends
+)
+def github_disclose_token(request):
+    # GitHub calls this API view when they have identified a string matching
+    # the regular expressions we provided them.
+    # Our job is to validate we're talking to github, check if the string contains
+    # valid credentials and, if they do, invalidate them and warn the owner
+
+    # The documentation for this process is at
+    # https://developer.github.com/partnerships/token-scanning/
+
+    body = request.body
+
+    # Thanks to the predicates, we know the headers we need are defined.
+    key_id = request.headers.get("GITHUB-PUBLIC-KEY-IDENTIFIER")
+    signature = request.headers.get("GITHUB-PUBLIC-KEY-SIGNATURE")
+
+    verifier = request.find_service(
+        IGitHubTokenScanningPayloadVerifyService, context=None
```

Comment:
Verifying the payload is costly and very complex:

- we call GitHub's meta API to extract key IDs and keys
- we check that the key ID is in our known keys
- we check the cryptographic signature

There are a lot of moving parts, and I imagined we may want to have simpler alternatives:
- Caching the meta API call in some sort of storage
- Actually storing those in the code directly, and avoid calling GitHub altogether if we want to avoid that for some reason
- Change the crypto impelementation
- ...

I though this part both was well contained, and had quite a few moving parts, so that's why I tried it this way. See https://github.com/pypa/warehouse/pull/7124#issuecomment-568281765 and below for what led me to that :)

Reply:
I agree that it makes sense to encapsulate this in a class or helper method, but a service is really for more generic things that we might want to someday swap out the underlying implementation.

The potential future changes you're describing mostly seem like things we'd just modify the service to support as necessary, rather than creating a new service entirely.

Like @dstufft mentioned before, a good example is our file storage. We want a generic interface to store files, and the ability to implement them as services that write to S3 or GCS or whatever, and the ability to swap them out via configuration whenever we need to.

Since this endpoint is very GitHub-specific, and the verifier is very GitHub-specific, I don't realistically see us ever replacing this with something else.

Reply:
Ok, I understand better now.